pkg/registry/generic/registry/store.go

```diff
@@ -532,82 +646,25 @@ func (e *Store) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 		return nil, kubeerr.NewInternalError(err)
 	}
 	pendingFinalizers := len(accessor.GetFinalizers()) != 0
-	var ignoreNotFound bool = false
-	var lastExisting runtime.Object = nil
-	// update the object
-	if graceful || (EnableGarbageCollector && (pendingFinalizers || shouldUpdateFinalizers(accessor, options))) {
-		out := e.NewFunc()
-		lastGraceful := int64(0)
-		err := e.Storage.GuaranteedUpdate(
-			ctx, key, out, false, &preconditions,
-			storage.SimpleUpdate(func(existing runtime.Object) (runtime.Object, error) {
-				graceful, pendingGraceful, err := rest.BeforeDelete(e.DeleteStrategy, ctx, existing, options)
-				if err != nil {
-					return nil, err
-				}
-				if pendingGraceful {
-					return nil, errAlreadyDeleting
-				}
-
-				// add/remove the orphan finalizer as the options dictates
-				existingAccessor, err := meta.Accessor(existing)
-				if err != nil {
-					return nil, err
-				}
-				updateFinalizers(existingAccessor, options)
-
-				if !graceful {
-					if EnableGarbageCollector {
-						// set the DeleteGracePeriods to 0 if the object has pendingFinalizers but not supporting graceful deletion
-						pendingFinalizers = len(existingAccessor.GetFinalizers()) != 0
-						if pendingFinalizers {
-							glog.V(6).Infof("update the DeletionTimestamp to \"now\" and GracePeriodSeconds to 0 for object %s, because it has pending finalizers", name)
-							err = notifyFinalizer(existing)
-							if err != nil {
-								return nil, err
-							}
-							return existing, nil
-						}
-					}
-					return nil, errDeleteNow
-				}
-				lastGraceful = *options.GracePeriodSeconds
-				lastExisting = existing
-				return existing, nil
-			}),
-		)
-		switch err {
-		case nil:
-			// if the object has pending finalizers, don't delete it, no matter
-			// whether the deletion is graceful, and no matter if the grace
-			// period is 0.
-			if EnableGarbageCollector && pendingFinalizers {
-				return out, nil
-			}
-			if lastGraceful > 0 {
-				return out, nil
-			}
-			// If we are here, the registry supports grace period mechanism and
-			// we are intentionally delete gracelessly. In this case, we may
-			// enter a race with other k8s components. If other component wins
-			// the race, the object will not be found, and we should tolerate
-			// the NotFound error. See
-			// https://github.com/kubernetes/kubernetes/issues/19403 for
-			// details.
-			ignoreNotFound = true
-			// exit the switch and delete immediately
-		case errDeleteNow:
-			// we've updated the object to have a zero grace period, or it's already at 0, so
-			// we should fall through and truly delete the object.
-		case errAlreadyDeleting:
-			return e.finalizeDelete(obj, true)
-		default:
-			return nil, storeerr.InterpretUpdateError(err, e.QualifiedResource, name)
+	var ignoreNotFound bool
+	var deleteImmediately bool = true
+	var lastExisting runtime.Object
+	out := e.NewFunc()
+	shouldUpdateFinalizers, _ := shouldUpdateFinalizers(accessor, options)
```

Comment:
> move inside else clause

Done

> do you really need to call this here? Can you not just call updateForGracefulDeletionAndFinalizers either way?

Probably I can remove the entire `if graceful || pendingFinalizers || shouldUpdateFinalizers`, given that we have the no-op update now, but I prefer to keep the redundancy for now, I can leave a Todo.